### PR TITLE
M-ACL: switch from UnsupportedAccess to AccessRestricted

### LINKED ISSUE #35177: M-ACL: Implement new error code "AccessRestricted"
A late M-ACL spec change requires returning a new error code, AccessRestricted, instead of UnsupportedAccess.  Rather than disrupt the near completion of the M-ACL PR (https://github.com/project-chip/connectedhomeip/pull/34932), this change will come immediately after that PR merges.

### PR DESCRIPTION
**Pull request by tleacmcsa**
Update AccessRestrictionProvider to return CHIP_ERROR_ACCESS_RESTRICTED_BY_ARL which is mapped to IM status AccessRestricted in the spec.

Fixes #35177

### CODE REVIEW

File: src/app/codegen-data-model-provider/CodegenDataModelProvider_Write.cpp [hunk 1]
```diff
@@ -287,10 +287,11 @@ DataModel::ActionReturnStatus CodegenDataModelProvider::WriteAttribute(const Dat
 
         if (err != CHIP_NO_ERROR)
         {
-            ReturnErrorCodeIf(err != CHIP_ERROR_ACCESS_DENIED, err);
+            ReturnErrorCodeIf(err != CHIP_ERROR_ACCESS_DENIED && err != CHIP_ERROR_ACCESS_RESTRICTED_BY_ARL, err);
 
             // TODO: when wildcard/group writes are supported, handle them to discard rather than fail with status
-            return Status::UnsupportedAccess;
+            return err == CHIP_ERROR_ACCESS_DENIED ? CHIP_IM_GLOBAL_STATUS(UnsupportedAccess)
+                                                   : CHIP_IM_GLOBAL_STATUS(AccessRestricted);
```
Comment (bzbarsky-apple) [inline src/app/codegen-data-model-provider/CodegenDataModelProvider_Write.cpp:294]: This should be returning Status, not CHIP_ERROR, from this location like it used to, I would think.

Comment (tleacmcsa) [reply src/app/codegen-data-model-provider/CodegenDataModelProvider_Write.cpp:294]: Oops, thanks.

Comment (tleacmcsa) [reply src/app/codegen-data-model-provider/CodegenDataModelProvider_Write.cpp]: Fixed.

File: src/app/util/ember-compatibility-functions.cpp [hunk 2]
```diff
@@ -701,9 +702,12 @@ CHIP_ERROR WriteSingleClusterData(const SubjectDescriptor & aSubjectDescriptor,
         }
         if (err != CHIP_NO_ERROR)
         {
-            ReturnErrorCodeIf(err != CHIP_ERROR_ACCESS_DENIED, err);
+            ReturnErrorCodeIf(err != CHIP_ERROR_ACCESS_DENIED && err != CHIP_ERROR_ACCESS_RESTRICTED_BY_ARL, err);
```
Comment (tcarmelveilleux) [inline src/app/util/ember-compatibility-functions.cpp:705]: Please parenthesize everywhere if possible:

```
err != CHIP_ERROR_ACCESS_DENIED && err != CHIP_ERROR_ACCESS_RESTRICTED_BY_ARL
```

becomes

```
(err != CHIP_ERROR_ACCESS_DENIED) && (err != CHIP_ERROR_ACCESS_RESTRICTED_BY_ARL)
```

**Review by tcarmelveilleux — APPROVED** [review]

**Review by hawk248 — APPROVED** [review]

File: src/app/util/ember-compatibility-functions.cpp (hunk 2 above)
Comment (tleacmcsa) [reply src/app/util/ember-compatibility-functions.cpp]: Done.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by chrisdecenzo — APPROVED** [review]

**Pull request merged**